### PR TITLE
GEOMESA-595 Ingest raster data in parallel

### PR DESCRIPTION
**Pull request by wjp22903**
Implement local version of ingesting raster files in parallel.

### CODE REVIEW

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala [hunk 1]
```diff
@@ -31,16 +31,18 @@ import org.joda.time.{DateTime, DateTimeZone}
...
+import scala.collection.parallel.ForkJoinTaskSupport
 import scala.util.Try
 
 object SimpleRasterIngest {
 
   def getReader(imageFile: File, imageType: String): AbstractGridCoverage2DReader = {
     imageType match {
-      case "tiff" => getTiffReader(imageFile)
-      case "dted" => getDtedReader(imageFile)
+      case TIFF => getTiffReader(imageFile)
```
Comment (aannex-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala]: This case match fails when files end with extension ".tif", please make this accept other known variations of the same file extension

Comment (wjp22903) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala]: Here is principle I used for file type:
- One type for one kind of images to simplify file type related process later on in the codes.
- To ingest Geotiff files under a directory, use format "-f tiff", it supports file extensions .tif and .tiff.
- To ingest DTED files under a directory, use format "-f dted", it supports file extensions .dt0. .dt1 and .dt2.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/Utils.scala [hunk 2]
```diff
@@ -61,8 +61,8 @@ object Utils {
   object Formats {
     val CSV     = "csv"
     val TSV     = "tsv"
-    val TIFF    = "tiff"
-    val DTED    = "dted"
+    val TIFF    = "GTiff"
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/Utils.scala]: I have a strong preference to spell out 'geotiff' unless there is a strong precedence for the abbreviation GTiff.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/AccumuloParams.scala [hunk 3]
```diff
@@ -83,20 +83,21 @@ class RasterParams extends AccumuloParams {
...
+  @Parameter(names = Array("-wt", "--write-threads"), description = "Threads for writing raster data")
   var writeThreads: Integer = null
 
-  @Parameter(names = Array("-qt", "--queryThreads"), description = "Threads for quering raster data")
+  @Parameter(names = Array("-qt", "--query-threads"), description = "Threads for quering raster data")
   var queryThreads: Integer = null
 
-  @Parameter(names = Array("-sh", "--shards"), description = "Number of shards to use for the storage tables (defaults to number of tservers)")
+  @Parameter(names = Array("-sh", "--shards"), description = "Number of shards to use for the storage tables " +
+    "(defaults to number of tservers)")
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/AccumuloParams.scala]: Good consistency!

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 4]
```diff
@@ -40,8 +40,8 @@ class IngestRasterCommand(parent: JCommander) extends Command with AccumuloPrope
       case TIFF | DTED =>
         ingest
       case _         =>
-        logger.error("Error: File format not supported for file " + params.file + ". Supported formats" +
-          "are tif, tiff, dt0, dt1 and dt2")
+        logger.error("Error: File format not supported for file " + params.file + ". Supported formats are GTiff and " +
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: (Picky) Long lines.  Ending at ... formats" will help.  Also, see note about GTiff.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/Utils.scala (hunk 2 above)
Comment (aannex-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/Utils.scala]: GTiff is the abbreviation used in GDAL for geotiff, not that we need to conform to that. Either way is fine.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 5]
```diff
@@ -52,7 +52,8 @@ class IngestRasterCommand(parent: JCommander) extends Command with AccumuloPrope
       IngestRasterParams.FILE_PATH -> Some(params.file),
       IngestRasterParams.FORMAT -> Some(Option(params.format).getOrElse(getFileExtension(params.file))),
       IngestRasterParams.RASTER_NAME -> Some(params.rasterName),
-      IngestRasterParams.TIME -> Option(params.timeStamp)
+      IngestRasterParams.TIME -> Option(params.timeStamp),
+      IngestRasterParams.PARLEVEL -> Option(params.parLevel.toString)
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: What happens if parLevel isn't provided? 

null.toString throws an NPE, right?

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 6]
```diff
@@ -100,8 +101,8 @@ object IngestRasterCommand {
 
   @Parameters(commandDescription = "Ingest a raster file of various formats into GeoMesa")
   class IngestRasterParameters extends CreateRasterParams {
-    @Parameter(names = Array("-fmt", "--format"), description = "Format of incoming raster data (tif | tiff | dt0 | " +
-      "dt1 | dt2) to override file extension recognition")
+    @Parameter(names = Array("-fmt", "--format"), description = "Format of incoming raster data (GTiff | DTED) to " +
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: Ditto for the GTiff note.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 7]
```diff
@@ -109,5 +110,9 @@ object IngestRasterCommand {
 
     @Parameter(names = Array("-tm", "--timestamp"), description = "Raster file to be ingested")
     var timeStamp: String = null
+
+    @Parameter(names = Array("-par", "--parallel-level"), description = "Maximum number of threads for ingesting " +
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: (Picky) Long lines: Break line after 'threads' or 'for'.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala (hunk 5 above)
Comment (wjp22903) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: params.parLevel has default value 1. So it's safe to call params.parLevel.toString.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala [hunk 8]
```diff
@@ -67,28 +69,47 @@ class SimpleRasterIngest(config: Map[String, Option[String]], cs: AccumuloCovera
...
+      (if (fileOrDir.isDirectory)
+         fileOrDir.listFiles(new FilenameFilter() {
+           override def accept(dir: File, name: String) =
+             getFileExtension(name).endsWith(fileType)
+         })
+       else Array(fileOrDir)).par
 
-    val envelope = rasterGrid.getEnvelope2D
+    files.tasksupport = new ForkJoinTaskSupport(new scala.concurrent.forkjoin.ForkJoinPool(parLevel))
+    files.foreach { file =>
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala]: (Picky) For readability, I'd suggest making this lamdba a separate function which takes a file and persists.

**Comment by jnh5y** [comment]
Jianping, looks good.  I added one more comment which can be addressed or left out.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala (hunk 8 above)
Comment (wjp22903) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/SimpleRasterIngest.scala]: Will do.

**Comment by ronq** [comment]
Merged into f_wcs_3 as commit e93f56645f8686a2d0e249e780035efd685ead70

**Pull request closed without merging**